### PR TITLE
Performance improvements within ExecutionHelper

### PR DESCRIPTION
**Pull request by Shane32**
Followup to #2165

### CODE REVIEW

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 1]
```diff
@@ -76,11 +76,16 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
         /// </summary>
         public static Variables GetVariableValues(Document document, ISchema schema, VariableDefinitions variableDefinitions, Inputs inputs)
         {
-            var variables = new Variables();
+            if ((variableDefinitions?.VariablesList?.Count ?? 0) == 0)
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:79]: Let's rename `VariablesList` to `List`.

File: src/GraphQL/Language/AST/Variables.cs [hunk 2]
```diff
@@ -55,6 +64,17 @@ public bool ValueFor(string name, out ArgumentValue value)
 
         /// <inheritdoc/>
         public override string ToString() => _variables?.Count > 0 ? $"Variables{{{string.Join(", ", _variables)}}}" : "Variables(Empty)";
+
+        /// <summary>
+        /// Returns a static instance that holds no variables.
```
Comment (sungam3r) [inline src/GraphQL/Language/AST/Variables.cs:69]: 👍

File: src/GraphQL/Language/AST/Variables.cs [hunk 3]
```diff
@@ -55,6 +64,17 @@ public bool ValueFor(string name, out ArgumentValue value)
...
+        /// </summary>
+        public static Variables None { get; } = new NoVariables();
+
+        private class NoVariables : Variables
```
Comment (sungam3r) [inline src/GraphQL/Language/AST/Variables.cs:73]:
```suggestion
        private sealed class NoVariables : Variables
```

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 4]
```diff
@@ -288,11 +336,27 @@ static object ParseValueObject(IInputObjectGraphType graphType, string variableN
...
+                //    foreach (var key2 in graphType.Fields.List)
+                //    {
+                //        if (key == key2.Name)
+                //        {
+                //            match = true;
+                //            break;
+                //        }
+                //    }
+
+                //    if (!match) (unknownFields ??= new List<string>(1)).Add(key);
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:353]: Nice implementation.

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 5]
```diff
@@ -248,9 +296,9 @@ static object ParseValueObject(IInputObjectGraphType graphType, string variableN
                 }
 
                 var newDictionary = new Dictionary<string, object>(dic.Count);
-                foreach (var field in graphType.Fields)
+                foreach (var field in graphType.Fields.List)
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:299]: 👍

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:353]: Oh - lol - I didn't finish it up obviously.  I was having trouble with tests, and then I guess I never went back to it 🤦‍♂️

Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:353]: I think it's all done; I'll remove the LINQ as long as it's working properly.

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 6]
```diff
@@ -288,11 +336,27 @@ static object ParseValueObject(IInputObjectGraphType graphType, string variableN
...
+                //            match = true;
+                //            break;
+                //        }
+                //    }
+
+                //    if (!match) (unknownFields ??= new List<string>(1)).Add(key);
+                //}
                 var unknownFields = dic.Keys
                     .Except(graphType.Fields.Select(f => f.Name))
                     .ToList();
```
Comment (Shane32) [inline src/GraphQL/Execution/ExecutionHelper.cs:357]:
```suggestion
                List<string> unknownFields = null;
                foreach (var key in dic.Keys)
                {
                    bool match = false;

                    foreach (var key2 in graphType.Fields.List)
                    {
                        if (key == key2.Name)
                        {
                            match = true;
                            break;
                        }
                    }

                    if (!match) (unknownFields ??= new List<string>(1)).Add(key);
                }
```

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 1 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs]: Done.

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 4 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs]: Done.

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 7]
```diff
@@ -131,6 +136,36 @@ public static Variables GetVariableValues(Document document, ISchema schema, Var
...
+            {
+                if (variableName.ChildName == null)
+                {
+                    Name = variableName.Name;
+                    Index = variableName.Index;
+                }
+                else
+                {
+                    Name = variableName;
+                    Index = default;
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:157]: Is `default` here `null` or `0` ?

File: src/GraphQL/Execution/ExecutionHelper.cs [hunk 8]
```diff
@@ -131,6 +136,36 @@ public static Variables GetVariableValues(Document document, ISchema schema, Var
...
+            public VariableName(VariableName variableName, string childName)
+            {
+                if (variableName.ChildName == null)
+                {
+                    Name = variableName.Name;
+                    Index = variableName.Index;
+                }
+                else
+                {
+                    Name = variableName;
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionHelper.cs:156]: So here will be concatenation via `ToString` on second (and deeper) level of depth?

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 7 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:157]: null

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 8 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:156]: yup

File: src/GraphQL/Execution/ExecutionHelper.cs (hunk 7 above)
Comment (Shane32) [reply src/GraphQL/Execution/ExecutionHelper.cs:157]: `Nullable<T>` is a struct so I used `default` but if you think `null` is better, I don't care.

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
foreach (var field in inputObjectGraphType.Fields)

**Comment by Shane32** [comment]
> foreach (var field in inputObjectGraphType.Fields) - https://github.com/graphql-dotnet/graphql-dotnet/pull/2165/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.md#r559019984

Fixed

**Pull request merged**